torch/csrc/distributed/autograd/utils.cpp

```diff
@@ -24,73 +24,83 @@ void addSendRpcBackward(
+      "Received tensors do not require grad, addRecvRpcBackward should not be called");
+  // Initialize autograd context if necessary.
+  auto& autogradContainer = DistAutogradContainer::getInstance();
+  DistAutogradContext& autogradContext =
+      autogradContainer.getOrCreateContext(autogradMetadata.autogradContextId);
+
+  // Attach the tensors as inputs to the autograd function.
+  auto grad_fn = std::make_shared<RecvRpcBackward>(
+      autogradMetadata, autogradContext, fromWorkerId);
+  for (auto& tensor : tensors) {
+    torch::autograd::set_history(tensor, grad_fn);
   }
-  return nullptr;
+
+  // Now update the autograd context with the necessary information.
+  autogradContext.addRecvFunction(grad_fn, autogradMetadata.autogradMessageId);
+  return &autogradContext;
 }
 
-std::shared_ptr<FutureMessage> sendMessageWithAutograd(
-    RpcAgent& agent,
-    const WorkerInfo& dst,
-    std::unique_ptr<RpcCommandBase> rpcCommand) {
+Message getMessageWithAutogradCheck(
+    const rpc::worker_id_t dstId,
+    torch::distributed::rpc::Message&& wrappedRpcMsg,
+    MessageType msgType) {
   auto& autogradContainer = DistAutogradContainer::getInstance();
-  if (autogradContainer.hasValidContext()) {
-    // Retrieve the appropriate context to modify.
-    auto& autogradContext = autogradContainer.currentContext();
-
-    // Wrap the original rpc with autograd information.
-    AutogradMetadata autogradMetadata(
-        autogradContext.context_id(), autogradContainer.newAutogradMessageId());
-    RpcWithAutograd rpcWithAutograd(
-        MessageType::MESSAGE_WITH_AUTOGRAD_REQ,
-        autogradMetadata,
-        std::move(rpcCommand));
-
-    // Record autograd information for 'send'.
-    addSendRpcBackward(
-        autogradContext, autogradMetadata, rpcWithAutograd.tensors());
-
-    return agent.send(dst, std::move(rpcWithAutograd).toMessage());
-  } else {
-    return agent.send(dst, std::move(*rpcCommand).toMessage());
+
+  if (!autogradContainer.hasValidContext() ||
+      !torch::autograd::compute_requires_grad(wrappedRpcMsg.tensors())) {
+    return std::move(wrappedRpcMsg);
   }
+
+  // Retrieve the appropriate context to modify.
+  auto& autogradContext = autogradContainer.currentContext();
+
+  // Wrap the original rpc with autograd information.
+  AutogradMetadata autogradMetadata(
+      autogradContext.contextId(), autogradContainer.newAutogradMessageId());
+  auto rpcWithAutograd = c10::guts::make_unique<RpcWithAutograd>(
+      RpcAgent::getDefaultRpcAgent()->getWorkerInfo().id_,
+      msgType,
+      autogradMetadata,
+      std::move(wrappedRpcMsg));
+
+  // Record autograd information for 'send'.
+  addSendRpcBackward(
+      autogradContext, autogradMetadata, rpcWithAutograd->tensors(), dstId);
+
+  return std::move(*rpcWithAutograd).toMessage();
+}
+
+std::shared_ptr<FutureMessage> sendMessage(
+    RpcAgent& agent,
+    const WorkerInfo& dst,
+    torch::distributed::rpc::Message&& wrappedRpcMsg,
```

Comment:
Can we pass in `wrappedRpc` in both sendMessage() and `getMessageWithAutogradCheck` with above and inside `getMessageWithAutogradCheck` call `toMessage` on the rpc? That seems cleaner from an API standpoint.

Reply:
in request_callback_impl.cpp, getMessageWithAutogradCheck is called, where we can only pass wrappedRpcResponse message to it.